### PR TITLE
Replication basics for MongoDB 

### PR DESCRIPTION
**Pull request by MiaAltieri**
This PR enables basic replication of the application MongoDB on Juju. It enables the user to start the application MongoDB with multiple units via `juju deploy ./*charm -n <number of desired replicas>`. It does **not** provide replication via `juju add-unit -n <number of desired replicas>`. It is however robust to deletion and a new primary will be nominated when the current one is deleted or otherwise fails

The following is NOT addressed in the PR, but is instead meant to be addressed in a future PR(s) if necessary
1. The replica set starts with the leader-unit as the primary. Unlike PostGres, MongoDB elects a new primary when the primary fails. This means we do **not** guarantee that axfter the original leader-unit fails that the new leader-unit is the primary. Do we want to guarantee this? (If we want the adminstrator to easily know which unit is the leader we could have an action called `get-primary` that returns the unit that contains the primary)
2. When a unit gets deleted it will still be present in `rs.stauts` and `rs.conf` this is so if that machine suddenly comes back up it can easily join the replicaset. But if it is not coming back into the replica set we may want to remove it. I asked this [on stackoverflow](https://stackoverflow.com/questions/70860622/mongodb-deleted-replicas-still-present-in-rs-conf-rs-status) I'm not fully sure what approach we should take, I think @balbir-thomas might have a good idea here on what should be done / what was done in the k8s charm

It is also important to note that some changes were made to `src/mongoserver.py` that add functionality to interacting wtih `mongod` as a single replica and as a whole replica, and also functionality to check whether the deployment is operating as a replica set

### CODE REVIEW

File: tests/unit/test_charm.py [hunk 1]
```diff
@@ -428,11 +452,230 @@ def test_add_mongodb_org_repository_already_added_skips(
...
+
+    @mock.patch("charm.MongodbOperatorCharm.app")
+    def test_mongodb_relation_joined_planned_units_does_not_match_joined_units(self, app):
+        rel = self.harness.charm.model.get_relation("mongodb")
+        self.harness.set_leader(True)
+        app.planned_units.return_value = 3
+
+        with self.assertLogs("charm", "DEBUG") as logs:
+            self.harness.charm.on.mongodb_relation_joined.emit(relation=rel)
+            self.harness.charm._add_mongodb_org_repository()
```
Comment (marceloneppel) [inline tests/unit/test_charm.py:480]: Is this call really needed at this test? It looks like it isn't, but I'm not sure.

File: src/charm.py [hunk 2]
```diff
@@ -79,29 +117,38 @@ def _on_start(self, event: ops.charm.StartEvent) -> None:
...
-            logger.debug("initialising replica set for the following IPs %s", self._unit_ips)
-            self._mongo.initialize_replica_set(self._unit_ips)
-            self._peers.data[self.app]["replica_set_hosts"] = json.dumps(self._unit_ips)
-        except (ConnectionFailure, ConfigurationError) as e:
-            logger.error("error initialising replica sets in _on_start: error: %s", str(e))
-            self.unit.status = BlockedStatus("failed to initialise replicaset")
+        # mongod is now active
+        self.unit.status = ActiveStatus()
+
+        # only leader should initialise the replica set
```
Comment (DnPlas) [inline src/charm.py:136]: We can probably let the `_mongodb_relation_joined()` event handler to manage the replicaset intialisation and let `_on_start()` manage the unit initialisation. I'm seeing this block of code is in both, so we might as well let just one event handler to do the task. I'm not sure if that works for MongoDB, though, this is just a suggestion.

File: src/charm.py [hunk 3]
```diff
@@ -205,6 +267,7 @@ def _config(self) -> dict:
             "root_password": "password",
             "security_key": "",
             "unit_ips": self._unit_ips,
+            "calling_unit_ip": str(self.model.get_binding(PEER).network.bind_address),
```
Comment (DnPlas) [inline src/charm.py:270]: Interesting. I will steal, I mean, leverage this for MySQL. (:

**Review by DnPlas — COMMENTED** [review]
Thanks @MiaAltieri , couple comments on the code, I didn't have the chance to see the tests, though.

File: tests/unit/test_charm.py [hunk 4]
```diff
@@ -168,131 +168,155 @@ def test_on_config_changed(self):
...
         service_resume.return_value = False
 
         with self.assertLogs("charm", "ERROR") as logs:
             self.harness.charm.on.start.emit()
+            service_resume.assert_called()
             self.assertIn("ERROR:charm:failed to enable mongod.service", logs.output)
         self.assertEqual(self.harness.charm.unit.status, BlockedStatus("couldn't start MongoDB"))
-        initialize_replica_set.assert_not_called()
+        _open_port_tcp.assert_not_called()
+        is_ready.assert_not_called()
```
Comment (marceloneppel) [inline tests/unit/test_charm.py:281]: It might also be worth checking that initialise_replica_set has not been called.

File: tests/unit/test_charm.py [hunk 5]
```diff
@@ -428,11 +452,230 @@ def test_add_mongodb_org_repository_already_added_skips(
...
+        # verify that we do not initialise replica set
+        initialise_replica_set.assert_called()
+
+    @patch_network_get(private_address="1.1.1.1")
+    @mock.patch("mongoserver.MongoDB.is_ready")
+    @mock.patch("mongoserver.MongoDB.initialise_replica_set")
+    @mock.patch("charm.MongodbOperatorCharm._open_port_tcp")
+    @mock.patch("charm.service_resume")
+    def test_initialise_replica_failure_leads_to_blocked_state(
+        self, service_resume, _open_port_tcp, initialise_replica_set, is_ready
```
Comment (marceloneppel) [inline tests/unit/test_charm.py:618]: `_open_tcp_port` parameter can be ommited with `_` as it's not being asserted for its call or something else (in this and in the following two functions).

**Review by marceloneppel — COMMENTED** [review]
Excellent work Mia! I left some comments to be evaluated.

File: src/charm.py [hunk 6]
```diff
@@ -31,12 +31,53 @@ def __init__(self, *args):
...
+        # do not initialise replica set until all peers have joined
+        if not self.app.planned_units() == len(self._peers.units) + 1:
+            logger.debug(
+                "planned units does not match joined units: %s != %s, will "
+                + "wait to initialise replica set until all planned units "
+                + "have joined",
+                str(self.app.planned_units()),
+                len(self.model.get_relation(PEER).units) + 1,
+            )
+            return
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:50]: That's what I initially thought too! But actually having a defer here leads to a forever wait. For example:

Say you have 3 units (`A`,`B`,`C`) (and `A` is the leader). When the leader observes the first `relation-joined` event this is because **one** of the units `B` joined, **but `C` has not joined** thus:

`self.app.planned_units() == len(self._peers.units) + 1` = `3 = 2` 

so the condition is not satisfied. If we defer we will **keep waiting**, and we will **not** observe `C` joining. In order to observe `C` we will need to wait for the **second** `relation-joined` event. Once the second `relation-joined` even fires we will observe `C` and: 

`self.app.planned_units() == len(self._peers.units) + 1` = `3 = 3`

And then will will be ready.

TL;DR: We cannot defer since we need to allow all `relation-joined` events to fire so we can observe all the peers. 

I hope that makes sense 😅 its kind of confusing, I'm also super happy to talk about this on MM. This was something that tripped me up for _awhile_

**Comment by MiaAltieri** [comment]
Repo now has a README.txt you will want to update this with respect to the added functionality here

File: src/charm.py [hunk 7]
```diff
@@ -31,12 +31,53 @@ def __init__(self, *args):
...
+            peer_config = self._config
+            peer_config["calling_unit_ip"] = peer_ip
+            mongo_peer = MongoDB(peer_config)
+            if not mongo_peer.is_ready(all_replicas=False):
+                logger.debug(
+                    "unit: %s is not ready, cannot initilise replica set "
+                    + "until all units are ready, defering on relation-joined",
+                    peer,
+                )
+                event.defer()
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:64]: So what `self._mongo.is_ready(all_replicas=True)` does is it checks if all the replicas in the replica set are available. (For this to return true the replica set **must** at first be initialised). Until it is, each replica will need to be checked individually which is why we check them in this way. Does that make sense?

You asking this makes me think I should change the variable `all_replicas` would that make things more clear? Perhaps something like `as_replica_set: bool` or `as_stand_alone: bool`?  I think `as_stand_alone` would make the most sense here, so it would be: `if not mongo_peer.is_ready(as_stand_alone=True):`

File: src/charm.py (hunk 3 above)
Comment (MiaAltieri) [reply src/charm.py:270]: Happy its useful! @marceloneppel also found it useful, perhaps we can add something to [unit](https://ops.readthedocs.io/en/latest/#ops.model.Unit) so we could do something like `self.unit.ip_address`. I'll try to remember to talk to @jnsgruk about this sometime this week

File: src/charm.py [hunk 8]
```diff
@@ -31,12 +31,53 @@ def __init__(self, *args):
...
+                return
+
+        # initialise replica set if not already done
+        if not self._mongo.is_replica_set():
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:68]: This is also confusing I apologise. The reason we don't block here is because it is possible that the application `MongoDB` could have been deployed with _one unit_ . If it is deployed as _one unit_ then it will initialise the replicaset in [_on_start](https://github.com/canonical/mongodb-operator/blob/41d005df7800f8e6a685e13a771f7b50371e4f7a/src/charm.py#L150-L151)

[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:68]: Hi deleted my first response because I misunderstood what you were saying. My apologies!

[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:68]: `_initialise_replica_set` will put it into a `BlockedStatus` if it isn't ready but I guess there is a need to handle an exception here so that way it doesn't execute the rest of the code after an initialisation failure. Thanks for catching this @DnPlas

File: src/charm.py [hunk 9]
```diff
@@ -31,12 +31,53 @@ def __init__(self, *args):
...
+            self._initialise_replica_set()
+
+        # verify replica set is ready
+        if not self._mongo.is_ready():
+            logger.debug("Replica set for units: %s, is not ready", self._unit_ips)
+            self.unit.status = BlockedStatus("failed to initialise replica set")
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:74]: Yes, so the `is_ready` here checks if the whole replicaset after initialisation is ready but the one above checks if each individual unit of MongoDB as the mongod (MongoDDB dameon running) I think I should change the optional variable name here so it makes more sense, let me tag you in that conversation. :)

File: src/charm.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:64]: @marceloneppel

File: tests/integration/test_charm.py [hunk 10]
```diff
@@ -75,8 +81,103 @@ async def test_database_is_up_as_replica_set(ops_test: OpsTest, unit_id: int) ->
...
+
+    print(client.primary)
+    client.close()
+
+    # verify that the primary is not None
+    assert primary is not None, "replica set with uri %s has no primary" % (
+        replica_set_uri)
+
+    # check that the primary is not one of the deleted units
+    print(primary[0])
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply tests/integration/test_charm.py:174]: oh god I've been caught. Yes let me delete that 🤦

File: src/charm.py [hunk 11]
```diff
@@ -79,29 +117,38 @@ def _on_start(self, event: ops.charm.StartEvent) -> None:
...
-        self.unit.status = ActiveStatus()
+        # do not initialise replica set until all peers have joined
+        if not self.app.planned_units() == len(self.model.get_relation(PEER).units) + 1:
+            logger.debug(
+                "planned units does not match joined units: %s != %s,"
+                + "waiting to initialise replica set",
+                str(self.app.planned_units()),
+                len(self.model.get_relation(PEER).units) + 1,
+            )
+            return
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:148]: If we `defer` here we will be in a situation where we forever wait for `self.app.planned_units() == len(self.model.get_relation(PEER).units) + 1` to be true. This is because the peer units won't be joined until the `relation-joined` hook is called. I have a more in depth explanation above for a similar situation in `relation-joined` let me tag you in it. :)

File: src/charm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:50]: @marceloneppel

File: tests/unit/test_charm.py (hunk 4 above)
Comment (MiaAltieri) [reply tests/unit/test_charm.py:281]: Agreed! Great eye Marcelo

File: tests/unit/test_charm.py (hunk 5 above)
Comment (MiaAltieri) [reply tests/unit/test_charm.py:618]: good eye!

File: src/charm.py (hunk 2 above)
Comment (DnPlas) [reply src/charm.py:136]: Ah, that makes sense.

File: src/charm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DnPlas) [reply src/charm.py:50]: I see. Since the deferred event is in the head of the queue, it won't update `self.model.get_relation(PEER)`, is that correct?

File: src/charm.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (marceloneppel) [reply src/charm.py:148]: Thanks for the explanation, Mia 😀. I understood this point. I'll just check your response for the last question from Daniela in the other comment.

File: src/charm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (marceloneppel) [reply src/charm.py:74]: Sorry Mia. I probably confused you with the word `above`. The call to `is ready` that I intended to comment about was the one from L72. The question is related to the fact that the `BlockedStatus` that is set inside the `if` clause is also set in L228, which is called in L69. I believe that there are two calls to set the same status. Maybe it's not the case if we can have `is ready` returning `False` after L224 not throwing an error. We can clarify this point better on MM.

File: src/charm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:50]: Yes exactly!

File: src/charm.py [hunk 12]
```diff
@@ -31,12 +31,53 @@ def __init__(self, *args):
...
+
+        # do not initialise replica set until all peers have joined
+        if not self.app.planned_units() == len(self._peers.units) + 1:
+            logger.debug(
+                "planned units does not match joined units: %s != %s, will "
+                + "wait to initialise replica set until all planned units "
+                + "have joined",
+                str(self.app.planned_units()),
+                len(self.model.get_relation(PEER).units) + 1,
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:48]: Going off of this I think it would be nice to have a something for the current number of units. So I added a property to the class for both of these. Let me know if that makes it confusing :)

File: src/charm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py]: Hmm I see what you're saying 🤔 . 

It's hard to say, I did this because I thought it _might_  be possible that `self._initialise_replica_set()` was successful but for some unforeseen reason `_mongo` might not be ready. `self._initialise_replica_set()` doesn't check if the replicaset as a whole is ready and the checks for `is_ready()` above only check if each unit has `mongod` running not the whole replicaset. So I think this check is still needed.

It might be best if we get the opinion of someone that knows MongoDB a bit better here. @balbirthomas what do you think of this?

[the remark this answers is not in the document]
Comment (marceloneppel) [reply src/charm.py]: Hi Mia! Thanks for the explanation and the details in our meeting today. I understood that makes sense to have this additional check for some situations where there is some time needed to the entire replica set to be ready.

File: src/charm.py [hunk 13]
```diff
@@ -31,12 +31,53 @@ def __init__(self, *args):
...
+            return
+
+        # do not initilise replica set until all peers have started mongod
+        for peer in self._peers.units:
+            peer_ip = str(self._peers.data[peer].get("private-address"))
+            peer_config = self._config
+            peer_config["calling_unit_ip"] = peer_ip
+            mongo_peer = MongoDB(peer_config)
+            if not mongo_peer.is_ready(all_replicas=False):
+                logger.debug(
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:60]: Thanks for pointing this out, I will make a different PR to update this across all string formatting cases. Thanks @DnPlas

File: tests/integration/test_charm.py [hunk 14]
```diff
@@ -75,8 +81,101 @@ async def test_database_is_up_as_replica_set(ops_test: OpsTest, unit_id: int) ->
...
+
+    This test will be implemented by Raul as part of his onboarding task.
+
+    Hint: multiple ways to do this, here are two but there are probably
+    other ways to implement it:
+    - check that each unit has the same replica set name
+    - create a URI for all units for the expected replica set name and see if
+      you can connect
+    """
+    pass
```
Comment (jnsgruk) [inline tests/integration/test_charm.py:95]: ?

File: tests/integration/test_charm.py [hunk 15]
```diff
@@ -85,3 +184,72 @@ def update_bind_ip(conf: str, ip_address: str) -> str:
...
+    # return None if primary is no longer in the replica set
+    if primary is not None and primary[0] not in valid_ips:
+        primary = None
+
+    return primary
+
+
+def is_zero(value):
+    """Return True if value is 0."""
+    return value == 0
```
Comment (jnsgruk) [inline tests/integration/test_charm.py:231]: ?

**Review by jnsgruk — CHANGES_REQUESTED** [review]

File: tests/integration/test_charm.py (hunk 14 above)
Comment (zmraul) [reply tests/integration/test_charm.py:95]: Maybe remove this test right now, and I can add it after the PR has been approved?

File: src/charm.py [hunk 16]
```diff
@@ -31,12 +31,61 @@ def __init__(self, *args):
...
+        self.framework.observe(
+            self.on.mongodb_relation_joined, self._mongodb_relation_joined)
         self.framework.observe(self.on.start, self._on_start)
 
+    def _mongodb_relation_joined(self, event) -> None:
+        if not self.unit.is_leader():
+            return
+
+        # do not initialise replica set until all peers have joined
+        if not self._number_of_expected_units == self._number_of_current_units:
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:43]: Hi @jnsgruk, good question! @marceloneppel and I thought that it would make things cleaner/easier to understand. If you think it makes things confusing let me know and I can revert :)

File: tests/integration/test_charm.py (hunk 15 above)
Comment (MiaAltieri) [reply tests/integration/test_charm.py:231]: @jnsgruk Same as above, calling function line 235

File: src/charm.py [hunk 17]
```diff
@@ -79,29 +125,43 @@ def _on_start(self, event: ops.charm.StartEvent) -> None:
...
-        self.unit.status = ActiveStatus()
+        # do not initialise replica set until all peers have joined
+        if not self._number_of_expected_units == self._number_of_current_units:
+            logger.debug(
+                "planned units does not match joined units: %s != %s,"
+                + "waiting to initialise replica set",
+                str(self._number_of_expected_units),
+                str(self._number_of_current_units),
+            )
+            return
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:157]: Yes I will opt for this instead and the remove the functions `_number_of_expected_units` & `_number_of_current_units`

File: src/charm.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:43]: Nevermind, I will opt for your suggestion below

File: tests/integration/test_charm.py [hunk 18]
```diff
@@ -85,3 +184,72 @@ def update_bind_ip(conf: str, ip_address: str) -> str:
...
+
+    Args:
+        ip_address: ip address of replica/unit
+    """
+    return "mongodb://{}:27017/".format(ip_address)
+
+
+def is_none(value):
+    """Return True if value is None."""
+    return value is None
```
[the remark this answers is not in the document]
Comment (jnsgruk) [reply tests/integration/test_charm.py:182]: You could use a lambda here - will ping you seperately

File: tests/integration/test_charm.py (hunk 15 above)
Comment (jnsgruk) [reply tests/integration/test_charm.py:211]: As above :)

File: src/charm.py [hunk 19]
```diff
@@ -31,12 +37,50 @@ def __init__(self, *args):
...
+            logger.debug("waiting to initialise replica set until all planned units have joined")
+            return
+
+        # do not initilise replica set until all peers have started mongod
+        for peer in self._peers.units:
+            peer_ip = str(self._peers.data[peer].get("private-address"))
+            peer_config = self._config
+            peer_config["calling_unit_ip"] = peer_ip
+            mongo_peer = MongoDB(peer_config)
+            if not mongo_peer.is_ready(as_stand_alone=True):
```
Comment (marceloneppel) [inline src/charm.py:58]:
```suggestion
            if not mongo_peer.is_ready(standalone=True):
```

File: src/charm.py [hunk 20]
```diff
@@ -79,29 +120,36 @@ def _on_start(self, event: ops.charm.StartEvent) -> None:
...
         try:
+            logger.debug("opening tcp port")
             self._open_port_tcp(self._port)
         except subprocess.CalledProcessError:
             self.unit.status = BlockedStatus("failed to open TCP port for MongoDB")
             return
 
-        if not self._mongo.is_ready():
+        # check if this unit's deployment of MongoDB is ready
+        if not self._mongo.is_ready(as_stand_alone=True):
```
Comment (marceloneppel) [inline src/charm.py:130]:
```suggestion
        if not self._mongo.is_ready(standalone=True):
```

**Review by marceloneppel — COMMENTED** [review]
Nice Mia. LGTM. I added two suggested changes to complement what Jon suggested about the `standalone` parameter.

File: src/charm.py (hunk 20 above)
Comment (MiaAltieri) [reply src/charm.py]: :face-palm: ty

**Review by jnsgruk — APPROVED** [review]

**Review by marceloneppel — APPROVED** [review]
